warn-stranded.lua

```diff
@@ -0,0 +1,204 @@
+
+   for index, id in pairs(currentIgnore['ints']) do
+      if id == unit.id then
+         print(id)
+         print(unit.id)
+         print('Unit Ignored')
+         return true, index
+      end
+   end
+
+   return false
+end
+
+local function toggleUnitIgnore(unit)
+   local currentIgnore = dfhack.persistent.get('warnStrandedIgnore')
+
+   if currentIgnore == nil then
+      currentIgnore = { key = 'warnStrandedIgnore', value = 'warnStrandedIgnore', ints = {} }
+      dfhack.printerr('nil, creating')
+      table.insert(currentIgnore['ints'], unit.id)
+   end
+
+   local ignored, index = unitIgnored(unit)
+
+   if ignored then
+      dfhack.printerr('toggle remove')
+      table.remove(currentIgnore['ints'], unit.id)
+   else
+      dfhack.printerr('toggle insert')
+      table.insert(currentIgnore['ints'], unit.id)
+   end
+
+   dfhack.persistent.save(currentIgnore)
+end
+
+function warning:initListChoices(units)
+   local choices = {}
+   for _, unit in pairs(units) do
+      table.insert(choices, { text = getUnitDescription(unit), unit = unit })
+   end
+   local list = self.subviews.list
+   list:setChoices(choices, 1)
+end
+
+function warning:updateBottom()
+   self.subviews.bottom_ui:setText(
+      {
+         { key = 'SELECT', text = ': Toggle ignore unit', on_activate = self:callback('onIgnore') }, ' ',
+         { key = 'CUSTOM_SHIFT_I', text = ': Ignore all', on_activate = self:callback('onIgnoreAll') }, ' ',
+         { key = 'CUSTOM_SHIFT_C', text = ': Clear all ignored', on_activate = self:callback('onClear') },
+      }
+   )
+end
+
+function warning:onIgnore()
+   local index, choice = self.subviews.list:getSelected()
+   local unit = choice.unit
+
+   toggleUnitIgnore(unit)
+end
+
+function warning:onIgnoreAll()
+   local currentIgnore = dfhack.persistent.get('warnStrandedIgnore')
+   local choices = self.subviews.list.getChoices()
+
+   if currentIgnore == nil then
+      currentIgnore = { key = 'warnStrandedIgnore', value = 'warnStrandedIgnore', ints = {} }
+   end
+
+   for _, choice in choices do
+      if not unitIgnored(unit) then
+         table.insert(currentIgnore['ints'], unit.id)
+      end
+   end
+
+   dfhack.persistent.save(currentIgnore)
+end
+
+function warning:onClear()
+   clear()
+end
+
+function warning:onDismiss()
+    view = nil
+end
+
+function doCheck()
+  local grouped = {}
+  local citizens = dfhack.units.getCitizens()
+
+  -- Pathability group calculation is from gui/pathable
+  for _, unit in pairs(citizens) do
+    local target = unit.pos
+    local block = dfhack.maps.getTileBlock(target)
+    local walkGroup = block and block.walkable[target.x % 16][target.y % 16] or 0
+    local groupTable = grouped[walkGroup]
+
+    if groupTable == nil then
+      grouped[walkGroup] = { unit }
+    else
+      table.insert(groupTable, unit)
+    end
+  end
+
+  local strandedUnits = {}
+
+
+  for _, units in pairs(grouped) do
+    if #units == 1 and not unitIgnored(units[1]) then
```

Comment:
do we only care if a single dwarf is stranded? what about closing a drawbridge and locking a group of foragers in the caverns? How about if we warn if we detect any bifurcation, and identify the smaller group as the "stranded"?

Reply:
I can work on that! The original issue specifically said a single dwarf but updating it should be easy

Reply:
For that matter if we really wanted we could have a configurable threshold.